src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+    printf("   Last possible lead-out start: %d:%02d.%02d (sector 0x%08X)\n",
+           (drive->rd_buf[20] << 8) | drive->rd_buf[21], drive->rd_buf[22],
+           drive->rd_buf[23], drive->media.last_lead_out);
+  }
+  /*
+      if (!drive->media.sstatus) {
+              drive->media.sessions--;
+              drive->media.tracks--;
+      }
+  */
+  return 0;
+}
+
+int determine_cd_type(drive_info *drive) {
+  //	unsigned char*	ATIP;
+  //	int		ATIP_len;
+  //	int		i;
+  int ratip = read_atip(drive);
+  if (ratip) {
+    if (!drive->silent)
+      printf("no ATIP found, assuming disc type: CD-ROM\n");
+    return DISC_CDROM; // CD-ROM
+                       //	} else {
+                       //		drive->ATIP_len += 4;
+  }
+  if (drive->media.ATIP_size < 8) {
+    if (!drive->silent)
+      printf("ATIP too small, assuming disc type: CD-ROM\n");
+    return DISC_CDROM; // CD-ROM
+  }
+  /*
+      printf("ATIP_len=%d\nATIP data:",drive->ATIP_len);
+      for (i=0; i< drive->ATIP_len; i++) printf("%4d",(drive->ATIP[i])&0xFF);
+      printf("\n");
+  */
+  // CD-RW?
+  int cdrw = !!(drive->media.ATIP[6] & 0x40);
+  int cd_subtype = (drive->media.ATIP[6] & 0x38);
+  if (cdrw) {
+    if (!drive->silent)
+      printf("disc type: CD-RW\n");
+    return (DISC_CDRW | cd_subtype);
+  }
+  if (!drive->silent)
+    printf("disc type: CD-R\n");
+  return (DISC_CDR | cd_subtype);
+}
+
+int read_mediaid_bd(drive_info *drive) {
+  int i;
+  unsigned int len;
+  memset(drive->media.MID, 0, 20);
+
+  for (int i = 0; i < speed_tbl_size; i++)
+    drive->parms.wr_speed_tbl_media[i] = -1;
+
+  drive->cmd[0] = MMC_READ_DVD_STRUCTURE;
+  drive->cmd[1] = 1; // media type = BD
+  drive->cmd[7] = 0x00;
+  drive->cmd[8] = 0;
+  drive->cmd[9] = 4;
+  drive->cmd[11] = 0;
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, 4))) {
+    if (!drive->silent)
+      sperror("READ_BD_STRUCTURE 00", drive->err);
+    return 1;
+  }
+  len = (drive->rd_buf[0] << 8 | drive->rd_buf[1]) + 2;
+  if (len > 128)
+    len = 128;
+  drive->cmd[0] = MMC_READ_DVD_STRUCTURE;
+  drive->cmd[1] = 1; // media type = BD
+  drive->cmd[7] = 0x00;
+  drive->cmd[8] = len >> 8;
+  drive->cmd[9] = len & 0xFF;
+  drive->cmd[11] = 0;
+  if ((drive->err = drive->cmd.transport(READ, drive->media.MID_raw, len))) {
+    if (!drive->silent)
+      sperror("READ_BD_STRUCTURE 00", drive->err);
+    return 1;
+  }
+  drive->media.MID_size = len - 4;
+
+  if (drive->media.MID_raw[4] != 'D' || drive->media.MID_raw[4 + 1] != 'I') {
+    printf("READ_BD_STRUCTURE: got some data, but not disc info\n");
+    return 2;
+  }
+  drive->media.disc_size = drive->media.MID_raw[4 + 11] >> 6;
+  drive->media.polarity = drive->media.MID_raw[4 + 14];
+
+  if (drive->media.type & DISC_BD_ROM) {
+    if (!drive->silent)
+      printf(COL_YEL "BD-ROM does not contain media ID" COL_NORM "\n");
+    return 0;
+  }
+  memcpy(drive->media.MID, drive->media.MID_raw + 4 + 100, 6);
+  i = strlen(drive->media.MID);
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)